### ISSUE TITLE
Unsure how to throw my own errors

### ISSUE THREAD

**Issue by VincentMeier** [graphql-java/graphql-spring-boot#24]
I'm new to both GraphQL and Spring boot, so bear with me please.

I'd like to throw my own errors, instead of getting the
```json
{
  "data": {
    "workflows": {
      "delete": null
    }
  },
  "errors": [
    {
      "message": "Internal Server Error(s) while executing query"
    }
  ]
}
```

I'm using the graphql-java-tools support and letting the starter auto configure everything. Just have the schema.graphqls in classpath and some Spring beans, that implement `com.coxautodev.graphql.tools.GraphQLResolver`.

I initially tried to code my resolver methods like so
```java
public String resolverMethod(String key) throws Exception {
    ...
    if (...some validation...) {
        ... some post validation ...
    } else {
        throw new graphql.GraphQLException("Validation failed");
    }
    ...
}
```
but in the end I get same error as before, and the Exception only appears in my server logs.

I read somewhere that it's the graphql servlet hiding the exception, but I don't know how to modify the servlet, since it's automatically created and configured by the Spring boot starter.

Long story short, I wanna know if I can return my own errors, while still using the Spring boot starter, or will I need to stop using it and configure everything manually from scratch?

**Comment by oliemansm** [comment]
You can override the default implementation of error handling by creating your own error handler. For example:
```java
public class CodedGraphQLErrorHandler extends DefaultGraphQLErrorHandler {

    @Override
    protected List<GraphQLError> filterGraphQLErrors(List<GraphQLError> errors) {
        return errors.stream()
                .map(this::codedError)
                .collect(Collectors.toList());
    }

    private GraphQLError codedError(GraphQLError error) {
        if (error.getErrorType() == ErrorType.DataFetchingException) {
            ExceptionWhileDataFetching exceptionWhileDataFetching = (ExceptionWhileDataFetching) error;
            Throwable exception = exceptionWhileDataFetching.getException();
            if (exception instanceof AccessDeniedException) {
                return new CodedGraphQLError(HttpStatus.FORBIDDEN, exception.getMessage());
            } else if (exception instanceof BadCredentialsException || exception instanceof PresetException) {
                return new CodedGraphQLError(HttpStatus.BAD_REQUEST, exception.getMessage());
            } else if (exception instanceof AuthenticationException) {
                return new CodedGraphQLError(HttpStatus.UNAUTHORIZED, exception.getMessage());
            }
        }
        return new CodedGraphQLError(HttpStatus.INTERNAL_SERVER_ERROR, error.getMessage());
    }
}


public class CodedGraphQLError extends GenericGraphQLError {

    private HttpStatus httpStatus;

    public CodedGraphQLError(HttpStatus httpStatus, String message) {
        super(message);

        this.httpStatus = httpStatus;
    }

    public int getCode() {
        return httpStatus.value();
    }

    public String getReason() {
        return httpStatus.getReasonPhrase();
    }
}
```
And then create your own instance of the GraphQLServlet like this
```java
@Configuration
public class GraphQLConfig {

    @Bean
    public GraphQLServlet graphQLServlet(GraphQLSchemaProvider schemaProvider, ExecutionStrategyProvider executionStrategyProvider) {
        return new SimpleGraphQLServlet(schemaProvider, executionStrategyProvider, null, null, new CodedGraphQLErrorHandler());
    }
}
```

**Comment by VincentMeier** [comment]
Judging by your code, this means I'll also need to build the GraphQLSchemaProvider and ExecutionStrategyProvider somehow, so that I can pass it to the servlet. Correct?

**Comment by oliemansm** [comment]
No, not necessarily because in the autoconfigurer those are created if they are missing. So you don't have to instantiate those, you can assume they are created.

**Comment by VincentMeier** [comment]
Yup. That did it. Major thanks :)

**Comment by apottere** [comment]
Just to clarify, the graphql autoconfiguration will wire in a `GraphQLErrorHandler` you provide to the context, so you don't need to create the `GraphQLServlet` yourself, just mark your `GraphQLErrorHandler` with `@Component`, or create a bean method that builds it.  https://github.com/graphql-java/graphql-spring-boot/blob/master/graphql-spring-boot-autoconfigure/src/main/java/com/oembedler/moon/graphql/boot/GraphQLWebAutoConfiguration.java#L79

**Comment by oliemansm** [comment]
@apottere Didn't realize that, nice

**Comment by gintautassulskus** [comment]
How do I return multiple validation errors for more that one field?

Should I create a custom exception that takes an array of messages and then process such an exception in my custom GraphQLErrorHandler?
